Linux-0.98/Yggdrasil-0.98.3/usr/src/usr.bin/utree.src/utree/doc/utree.1
'''\"
'''\"   utree.man, utree 3.03-um manual page
'''\"   klin, Fri Mar  6 11:31:02 1992
'''\"
'''\"   [x]roff -man utree.man
'''\"   groff   -man [-Tdevice] utree.man
'''\"
'''\"   SCCSID = @(#) utree.man 3.03-um Mar  6 1992
'''\"
.TH UTREE 1L "UTREE Version 3.03-um" "March 6 1992"
.SH NAME
.LP
\fButree\fR \- Screen oriented filesystem browser and utility
.SH SYNOPSIS
.LP
utree [options] [rootdirectory]
.SH DESCRIPTION
.LP
The screen oriented filesystem browser and utility
\fButree\fR descends the directory hierarchy
rooted in your home directory,
defined in a tree list file if given,
or rooted in \fBrootdirectory\fR if given,
and displays the directory tree on the so called \fItree screen\fR.
On the \fItree screen\fR \fButree\fR lets you walk thru the tree,
execute some default and user defined commands on it
and change to any directory in the tree.
.br
If you have changed to a directory \fButree\fR displays all files
in this directory on the so called \fIfile screen\fR.
On the \fIfile screen\fR \fButree\fR lets you walk thru the file list
and execute some default and user defined commands on files.
.SH OPTIONS
.LP
The following command line options are interpreted by \fButree\fR:
.sp
.TP 16
\fB-L\fR
Follow symbolic links to directories
.TP
\fB-S\fR
Ignore default minimal screen size of 80 columns and 24 lines
.TP
\fB-V\fR
Display \fButree\fR version and copyright notice
.TP
\fB-a\fR
Read in all directories,
including those that begin with a dot (.),
which are normally skipped
.TP
\fB-b\fR
Suppress ringing of the bell (useful for visual bells)
.TP
\fB-c\fR
Don't display and update a clock every second
.TP
\fB-d var=[val]\fR
Set the variable \fBvar\fR to value \fBval\fR
or unset the variable \fBvar\fR
.TP
\fB-d typ:[cmd]\fR
Define the command \fBcmd\fR for the filetype \fBtyp\fR
or undefine any command for the filetype \fBtyp\fR
.TP
\fB-f lst\fR
Build directory tree from list file \fBlst\fR
.TP
\fB-g\fR
Don't use graphical characters
.TP
\fB-h\fR
Display usage and some help about options
.TP
\fB-i ind\fR
Set the tree level indention to \fBind\fR columns (3..9)
.TP
\fB-n\fR
Avoid scanning the tree for changes i.e. after shell escape
.TP
\fB-l lev\fR
Build the tree up to level \fBlev\fR for quicker startup
.TP
\fB-o\fR
Omit saving changes in variables and filetype command definitions,
command history and key bindings
.TP
\fB-p lin\fR
Use \fBlin\fR lines for displaying filenames of the current directory
on the \fItree screen\fR (default: 3)
.TP
\fB-q\fR
Build the tree up to level 2 (like -l 2)
.TP
\fB-r\fR
Build the tree scanning and reading the file system
instead of reading a tree list file
.TP
\fB-s\fR
Don't use hardware scrolling
.TP
\fB-u\fR
Read and update all file lists after building
the tree from a tree list file
.TP
\fB-v mod\fR
Set using video attributes to mode \fBmod\fR.
\fBmod\fR may be \fB2\fR for all possible,
\fB1\fR for bold and underline
and \fB0\fR for none video attributes.
.TP
\fB-w\fR
Suppress warnings about unreadable directories
.TP
\fB-x cmd\fR
Use and execute the string \fBcmd\fR as initial input at startup.
The string \fBcmd\fR is a simple sequence of \fButree\fR commands
.PP
The boolean options \fBbcgnosw\fR and the
numerical options \fBipv\fR may also be pre-set in
the environment variable \fBUTREE\fR.
E.g. if you want to suppress ringing of the bell and
displaying and updating the clock and if you prefer a tree indention
of 4 columns set the environment variable \fBUTREE\fR to '\fBbci4\fR'
for this reason.
.br
Most of the options correspond to \fButree\fR
variables and therefore they may also be set or unset
with the variable command (\fB=\fR).
See the sections \fIglobal commands\fR and \fIvariables\fR below.
.SH UTREE STARTUP
.LP
If \fButree\fR is called without the optional command line parameter
\fBrootdirectory\fR or this directory is rooted in your
home directory \fButree\fR tries to build up the directory tree
reading a file \fB.utreelist\fR in your home directory,
which contains a list of your directory tree created
by a previous \fButree\fR session
or by the additional shell script \fButree.mklist\fR called before.
If the command line option \fB-f lst\fR is given
\fButree\fR builds up the tree from this list file \fBlst\fR
which may be created with the denoted shell script \fButree.mklist\fR
or a command like \fBfind\fR.
Entries in such list files have to start with the
directory separator / (slash) in the first column of the line.
All other entries are ignored and skipped.
.sp
If building the tree from a list file
the file lists of directories are created and read in on demand only,
which means when a directory becomes the current directory
by moving the cursor to this directory.
This speeds up the start of \fButree\fR because there is no need
for scanning and reading the filesystem for subdirectories
what may take some time for larger filesystems.
.sp
In all others cases,
if no list file is given with the
command line option \fB-f lst\fR,
\fBrootdirectory\fR is rooted in your home directory
and the file \fB.utreelist\fR is not found,
or the command line option \fB-r\fR is given
\fButree\fR builds up the tree by
scanning and reading the filesystem recursively.
You can speed up the start with the commandline option
\fB-l lev\fR which causes \fButree\fR to build the
initial directory tree only up to level \fBlev\fR similar to
the option \fB-l\fR of the well known \fBfind\fR command.
At runtime some commands let you expand the directory
tree later and inspect directories and subtrees
not visible at startup (see below).
.SH SCREEN LAYOUT
.LP
The screen is divided into three regions.
The top screen line,
the so called \fIecho line\fR is for displaying
messages and entering input.
The second screen line,
the so called \fIhelp line\fR is for displaying
help messages and the \fButree\fR default
or user defined menu lines.
.br
The rest of the screen forms a window on the directory tree
on the \fItree screen\fR or
on the file list of the current directory
on the \fIfile screen\fR.
Or he is used for displaying variable or filetype commands settings,
for displaying help pages and displaying command outputs.
.SH UTREE SCREENS AND MENUS
.LP
\fBUtree\fR knows the following screens and menus:
.sp
.TP 20
\fBtree screen\fR
This is the initial screen displayed when \fButree\fR is started
and has built up the directory tree.
The \fItree screen\fR forms a window on the directory tree.
The current directory on which most commands are
working is highlighted and the last
screen lines are used to display the first files
of the file list of the current directory.
You may move the window over the directory tree,
enlarge or shrink the directory tree window
or walk thru the directory tree and execute
default and user defined commands on directories and subtrees.
On the help line the most important commands available
for directories or subtrees on
the \fItree screen\fR are shown in the default \fItree menu\fR.
The tree menu shows the names of the
the \fItree screen\fR commands,
the keystrokes to invoke a command are marked with uppercase letters.
You may switch the menu to a user defined menu line
displaying user defined commands for the \fItree screen\fR.
For further information about all \fItree screen\fR commands
see the sections \fIglobal commands\fR,
\fItree screen commands\fR and \fIvariables\fR of this manual page.
.sp
.TP
\fBfile screen\fR
The \fIfile screen\fR forms a window on all files of the
current directory.
The current file on which most commands are
working is highlighted.
Yoy may move the window over the file list
or walk thru the file list and execute
default and user defined commands on files.
On the help line the most important commands
for files available on
the \fIfile screen\fR are shown in the default \fIfile menu\fR.
The file menu displays the names of the
the \fIfile screen\fR commands,
the keystrokes to invoke a command are marked with uppercase letters.
You may switch the menu to a user define menu line
displaying user defined commands for the \fIfile screen\fR.
For further information about all \fIfile screen\fR commands
see the sections \fIglobal commands\fR,
\fIfile screen commands\fR and \fIvariables\fR of this manual page.
.sp
.TP
\fBhelp screen\fR
If you have switched to the \fIhelp screen\fR with
the help command (\fBh\fR)
on the help line the
\fIhelp menu\fR is displayed and you may
select help about interesting topics.
The help menu displays the names of all
available help topics,
the keystrokes to select help about a topic
are marked with uppercase letters.
For more information see the section \fIhelp pages\fR below.
.sp
.TP
\fBvariables screen\fR
If you have switched to the \fIvariables\fR screen with
the variables command (\fB=\fR)
all variables and their settings are displayed
and you may set or unset any of the variables.
On the help line is shown how to set or unset a variable
in a short form.
All changes in variables definitions are
saved to the file \fB.utree\fR in your home directory
if the variable \fBAUTOSAVE\fR is set.
For more information see the section \fIvariables\fR below.
.sp
.TP
\fBcommands screen\fR
If you have switched to the filetype \fIcommands screen\fR with
the filetype commands command (\fB:\fR)
all filetype commands and their settings are displayed
and you may set or unset any of the filetype commands.
On the help line is shown how to set or
unset a filetype command in a short form.
All changes in filetype command definitions are
saved to the file \fB.utree\fR in your home directory
if the variable \fBAUTOSAVE\fR is set.
For more information see the section \fIfiletype commands\fR below.
.sp
.TP
\fBshell screen\fR
If you have switched to the \fIshell screen\fR with
the shell command (\fB!\fR) for executing commands
not supported directly by \fButree\fR
all previously entered commands saved in a so called
history list are displayed on the \fIshell screen\fR.
You may get any command from the history list
into the line editor for editing and execution
or enter and execute a new command.
All commands up to a maximal number defined in
the variable \fBHISTSIZE\fR are saved in the
history list.
Before leaving \fButree\fR all saved commands
in the history list are saved to a
history file \fB.utreehist\fR in your home directory
if the variable \fBAUTOSAVE\fR is set.
At startup this file is searched for and read in
if found.
For more information see also the section \fIvariables\fR below.
.sp
.TP
\fBbindings screen\fR
If you have switched to the \fIbindings screen\fR
with the bindings command (\fB|\fR) all
currently defined key bindings and their meaning are
displayed and you may bind any key to an appropriate
utree command or to insert a string into the
input buffer.
At startup a file \fB.utree-TERM\fR in your home directory
or a file \fButree-TERM\fR in a global startup directory
is read in if found containing key bindings for
the terminal type defined in the environment variable \fBTERM\fR.
All changes in key bindings you have done on the
\fIbindings screen\fR at runtime are saved to \fB.utree-TERM\fR
if the variable \fBAUTOSAVE\fR is set.
For more information see also the section \fBkey bindings\fR
below.
.sp
.TP
\fBstatus screen\fR
The status command (\fBs\fR) displays all information
available about a file or directory on the \fIstatus screen\fR.
Here you may change the ownership, group membership
or access rights of a file or directory.
On BSD systems not all of the denoted changes may be allowed
for normal users.
.SH KEY NAMING CONVENTIONS AND DEFAULT KEY BINDINGS
.LP
All \fButree\fR commands are simple single letter commands or
control sequences.
The default or user defined commands therefore
are invoked with a single keystroke or a combination
of the <\fICONTROL\fR> key with
another key.
The naming conventions in the following manual sections
for \fButree\fR commands invoked by a keystroke are:
.sp
.TP  20
\fBkey\fR
means hit this <\fIkey\fR> only
.TP
\fBC-key\fR
means hold down the <\fICONTROL\fR>-key and hit <\fIkey\fR>
.PP
To permit rebinding of pre-defined keys or
binding functions keys to \fButree\fR commands
all control sequences have special names.
All key or functions names and their default bindings
are list in the table below
.TP 20
\fBSELECT\fR
CR, NL
.TP
\fBFORWARD\fR
C-f
.TP
\fBBACKWARD\fR
C-b
.TP
\fBNEXT\fR
C-n
.TP
\fBPREVIOUS\fR
C-p
.TP
\fBNEXTPAGE\fR
C-v
.TP
\fBPREVPAGE\fR
C-w
.TP
\fBBEGIN\fR
C-a
.TP
\fBEND\fR
C-e
.TP
\fBUP\fR
C-u
.TP
\fBDOWN\fR
C-d
.TP
\fBINSERT\fR
C-o
.TP
\fBDELETE\fR
BS
.TP
\fBKILL\fR
C-k
.TP
\fBSETMARK\fR
C-@
.TP
\fBGOTOMARK\fR
C-g
.TP
\fBGOTOTAG\fR
C-t,TAB
.TP
\fBHELP\fR
C-r
.TP
\fBREFRESH\fR
C-l
.TP
\fBCANCEL\fR
C-x
.TP
\fBBREAK\fR
C-c,C-y
.TP
\fBEXIT\fR
C-z
.PP
In the following sections of this manual page the
default key bindings are used instead of the
names or functions.
.PP
Some function keys are supported by \fButree\fR,
i.e. the four arrow or cursor keys,
and are bound to appropriate functions.
See the sections \fIkey bindings\fR and \fIfunction keys\fR
below for more details.
.SH HELP PAGES
.LP
If the \fButree\fR help pages contained in the file \fButree.help\fR
are accessible in a directory defined at compile time or defined in
the environment variable \fBUTLIB\fR you can get help
on all screens or from within the line editor
with the help command (\fBh\fR or \fB?\fR) or the help key (\fBC-r\fR).
After displaying help about your current context,
i.e. help about tree commands if you are on the \fItree screen\fR,
you can switch to the help menu and
select help about all topics with a single keystroke.
All \fIhelp screen\fR commmands and the menu items of the help menu
displayed on the help line are:
.sp
.TP 16
\fBh\fR
(Help) About the help pages and the help menu
.TP
\fBa\fR
(About) Information about \fButree\fR and key naming conventions
.TP
\fBu\fR
(Usage) Description of \fButree\fR usage and commandline options
.TP
\fBg\fR
(Global) Global commands common for the \fItree\fR and the \fIfile screen\fR
.TP
\fBt\fR
(Tree) Commands for the \fItree screen\fR
.TP
\fBf\fR
(File) Commands for the \fIfile screen\fR
.TP
\fBe\fR
(Edit) Commands of the builtin line editor
.TP
\fBv\fR
(Vars) \fBUtree\fR variables and variable definition
.TP
\fBc\fR
(Cmds) Filetype commands and filetype command definition
.TP
\fBl\fR
(Line) Line format for user defined tree, file and filetype commands.
.TP
\fBk\fR
(Keys) Function keys used by \fButree\fR
.TP
\fBp\fR
(Patterns) File pattern matching, filename, modification time and
file size patterns
.TP
\fBq\fR
(Quit)
Leave \fIhelp screen\fR
.PP
The help pages contain in short form most information given
in the sections of this manual page.
.SH GLOBAL COMMANDS
.LP
The following commands are common for the \fItree\fR and \fIfile screen\fR
They can be given in lowercase or uppercase letters:
.sp
.TP 20
\fBC-z\fR
Exit \fButree\fR from all screens
.TP
\fBC-c,C-y\fR
Cancel or break current command or input
.TP
\fBC-l\fR
Redisplay the current screen or the input line
.TP
\fBTAB,C-t\fR
Move to the next tagged file or the next directory
containing tagged files
.TP
\fBh,?,C-r\fR
Display help pages and switch to the \fIhelp screen\fR
and the help menu
.TP
\fB@,C-@\fR
Mark the current directory or file
.TP
\fB#,C-g\fR
Goto to a previously marked directory or file
.TP
\fBa\fR
Display \fButree\fR version and copyright notice
.TP
\fBd\fR
Display current date and time
.TP
\fBj\fR
Move to the next directory or file (for vi fans)
.TP
\fBk\fR
Move to the previous directory or file (for vi fans)
.TP
\fBn\fR
Change sort criteria from lexical order to modification time order
or vice versa and resort files in the file list
.TP
\fBt\fR
Tag files matching a file pattern for further processing
.TP
\fBu\fR
Untag files
.TP
\fBw\fR
Display full pathname of the current directory
.TP
\fBz\fR
Zoom files from filelist matching a file pattern
.TP
\fB=\fR
Switch to the \fIvariables screen\fR,
display and set or unset variables
.TP
\fB:\fR
Switch to the \fIcommands screen\fR,
display and set or unset
filetype commands
.TP
\fB|\fR
Switch to the \fIbindings screen\fR,
display all key bindings
and bind or rebind keys
.TP
\fB!\fR
Switch to the \fIshell screen\fR, display all commands
from the shell command history list, enter and execute
commands not supported directly from \fButree\fR.
Before a given command is executed the command line
is searched for some sprintf like format characters
lead in by a percent sign (%) which are expanded.
See the section \fIline formats\fR for more
information
.TP
\fB$\fR
Escape to an interactive shell
.SH TREE SCREEN COMMANDS
.LP
All commands on the \fItree screen\fR can be given in
lowercase or uppercase letters with the meaning denoted below.
Commands given in lowercase letters affect the current directory only.
Commands given in uppercase letters
indicated by an uppercase letter in the table below
affect the subtree rooted in the current directory
or all tagged files in the subtree rooted in the current directory.
The \fItree screen\fR commmands and the menu items of the
default tree menu displayed on the help line are:
.sp
.TP 20
\fB>,CR,NL,SP,>\fR
Change to the \fIfile screen\fR of the current directory
.TP
\fB<\fR
Change to the \fIfile screen\fR of the parent directory
.TP
\fBC-n\fR
Move to the next directory
.TP
\fBC-p\fR
Move to the previous directory
.TP
\fBC-f\fR
Move to the next directory on same level as the current directory
.TP
\fBC-b\fR
Move to the previous directory on same level as the current directory
.TP
\fBC-v\fR
Move one page forward
.TP
\fBC-w\fR
Move one page backward
.TP
\fBC-a\fR
Move to the beginning of the directory tree
.TP
\fBC-e\fR
Move to the end of the directory tree
.TP
\fBTAB,C-t\fR
Move to next the directory containing tagged files
.TP
\fBC-u\fR
Scroll up one line the directory tree
.TP
\fBC-d\fR
Scroll down one line the directory tree
.TP
\fB@,C-@\fR
Mark the current directory
.TP
\fB#,C-g\fR
Move to a previously marked directory
.TP
\fBh,?\fR
(Help) Help about \fItree screen\fR commands
.TP
\fBb,B\fR
(Backup) Backup the current directory or tree or
backup all tagged files in the subtree
.TP
\fBc,C\fR
(Chdir) Move to a directory or copy all tagged files
in the subtree. Before you move to a directory you are requested
for the name of this directory before.

Linux-0.98/Yggdrasil-0.98.3/usr/src/usr.bin/utree.src/utree/src/utree.prlist.c
/*
 *      UTREE.PRLIST.C
 *      Filter for utree 3.02-um tree list files for different devices.
 *      3.02-um klin, Sun Nov 10 13:13:35 1991
 *              klin, Sun Nov 24 12:07:16 1991, XENIX graphset handling
 *      3.03-um klin, Sun Feb 16 15:34:37 1992, Minor changes
 *            a klin, Sun Mar 15 19:08:25 1992, Minor changes for AIX 3.2
 *
 *      Copyright (c) 1991/92 by Peter Klingebiel & UNIX Magazin Muenchen.
 *      For copying and distribution information see the file COPYRIGHT.
 */
static char sccsid[] = "@(#) utree 3.03a-um (klin) Mrz 15 1992 utree.prlist.c";

/*
 *      Usage:  utree.prlist [-Vhr] [-d dev] [-f fnt] [i ind] [-s siz] listfile
 *      Options:        -V      Display program version.
 *                      -T dev  or
 *                      -d dev  Output for device dev.
 *                              Supported devices are:
 *                              ascii   ascii meta characters.
 *                              850     printers with ibm pc850 charset.
 *                              ps      postscript.
 *                              term    terminal (default).
 *                      -f fnt  Font to use (postscript only)
 *                      -h      Display usage and help.
 *                      -i ind  Set tree indention to ind.
 *                      -s siz  Font size (postscript only)
 *      Remarks:        Formatted utree tree list files can be created
 *                      with the write-tree command (o) and option list (l)
 *                      from tree-screen.
 *      Directory:      /usr/local/bin
 *      Environment:    $TERM
 *      Defines:        AIX     for IBM AIX systems
 *                      BSD     for BSD systems
 *                      XENIX   for XENIX systems
 *      Libraries:      -lm -lcurses (SYSV), -ltermcap or -ltermlib (BSD)
 */

static char usage[] = "@(#) Usage: utree.prlist [-Vr][-d dev][-f fnt][-i ind][-s siz] listfile";
static char pvers[] = "UTREE 3.03-um";

#include <stdio.h>
#include <signal.h>

#define MAXLEN  512             /* Buffer length                */
#define DEFIND  6               /* Default indention            */
#define MININD  3               /* Minimal indention            */
#define MAXIND  9               /* Maxinal indention            */
#define OPTIONS "T:Vd:f:hi:rs:" /* Optionstring for getopt()    */
#define UNULL   ((unsigned char *) 0)

static char *pname;
static char *title = "Tree";
static char root[MAXLEN];

/* The generic graphical character set                          */
static unsigned char *grafset = UNULL;
#define H grafset[0]            /* Horizontal bar               */
#define V grafset[1]            /* Vertical bar                 */
#define U grafset[2]            /* Upper left corner            */
#define L grafset[3]            /* Lower left corner            */
#define T grafset[4]            /* Left tee                     */
#define I grafset[5]            /* Indention                    */
#define S grafset[6]            /* Space                        */

/* The fix coded character sets for ascii, acsc, ibm and ps     */
/*                                    H   V   U   L   T   I   S */
static unsigned char asciiset[7] = { '-','|','+','+','|',' ',' ' };
static unsigned char acscset[7]  = { '-','|','+','+','|',' ',' ' };
static unsigned char ibmset[7]   = { 196,179,218,192,195,' ',' ' };
static unsigned char psset[7]    = { 'H','V','U','L','T','I','S' };

static int indent = DEFIND;     /* Indention                    */
static int rflag  = 1;          /* Print rootdirectory flag     */
static int isps   = 0;          /* Postscript output            */

static unsigned char *acsc  = UNULL;    /* Alternate char set   */
static unsigned char *enacs = UNULL;    /* Enable alt char set  */
static unsigned char *smacs = UNULL;    /* Alt char set on      */
static unsigned char *rmacs = UNULL;    /* Alt char set off     */
#define ACSCON  if(acsc && smacs) tputs(smacs, 1, putchar)
#define ACSCOFF if(acsc && rmacs) tputs(rmacs, 1, putchar)
#undef  putchar

/* Postscript definitions and variables                         */
#define PSFONT  "Courier-Bold"  /* Default postscript font      */
#define PSSIZE   10.0           /* Default postscript font size */
#define PSMIN    50.0           /* Lower y position             */
#define PSMAX   780.0           /* Upper y position             */
#define PSIND    85.0           /* Most left x position         */
#define PSWID   475.0           /* Page header width            */
#define PSEOF   (0x04)          /* C-d: EOF for postscript      */
static int psp = 0;             /* Page counter                 */
static double psy;              /* Current y position           */
static double pps = PSSIZE;     /* Font size                    */
static double pvs = PSSIZE+2;   /* Vertical space               */
static char *pft = PSFONT;      /* Font                         */

extern double atof();
#ifdef  BSD
#define strrchr rindex
extern char *rindex();
#else   /* !BSD */
extern char *strrchr();
#endif  /* BSD */

/*
 *      Output one character
 */
static int putchar(c)
  int c;
{
  putc((unsigned char) c, stdout);
  return(c);

} /* putchar() */

/*
 *      Catch signals SIGINT, SIGQUIT and SIGTERM
 */
static int catchsig(sig)
  int sig;
{
  ACSCOFF;
  if(sig == SIGQUIT)
    abort();
  exit(1);

} /* catchsig() */

/*
 *      Init terminal and graphical character set for terminal
 */
static void initterm(f)
  int f;
{
  static char termbuf[1024];
  char termcap[1024], *term;
  char *tp = termbuf;
  extern char *getenv();
  extern unsigned char *tgetstr();

  grafset = asciiset;           /* Default character set */
  if(f && !isatty(fileno(stdout)))
    return;
  else if((term = getenv("TERM")) && tgetent(termcap, term) > 0) {
#ifdef  AIX
    if(acsc = tgetstr("bx", &tp)) {
      grafset = acscset;
      if( !(smacs = tgetstr("as", &tp)))
     smacs = tgetstr("f1", &tp);     /* AIX 3.2: font1 instead of smacs */
      if( !(rmacs = tgetstr("ae", &tp)))
     rmacs = tgetstr("f0", &tp);     /* AIX 3.2: font0 instead of rmacs */
      U = acsc[0];
      L = acsc[5];
      H = acsc[1];
      T = acsc[9];
      V = acsc[3];
    }
#else   /* !AIX */
    if(acsc = tgetstr("ac", &tp)) {
      grafset = acscset;
      if(enacs = tgetstr("eA", &tp))
       tputs(enacs, 1, putchar);
      smacs = tgetstr("as", &tp);
      rmacs = tgetstr("ae", &tp);
      do
       switch(*acsc) {
         default:              /* Skip */
           ++acsc; break;
         case 'l':             /* Upper left corner */
           U = *++acsc; break;
         case 'm':             /* Lower left corner */
           L = *++acsc; break;
         case 'q':             /* Horizontal bar */
           H = *++acsc; break;
         case 't':             /* Left tee */
           T = *++acsc; break;
         case 'x':             /* Vertical bar */
           V = *++acsc; break;
       }
      while(*acsc && *++acsc);
    }
# ifdef XENIX
    else {
      unsigned char *gp;

      grafset = acscset;
      acsc  = acscset;          /* Use as flag */
      smacs = tgetstr("GS", &tp);
      rmacs = tgetstr("GE", &tp);
      if(gp = tgetstr("G2", &tp))
       U = *gp;                /* Upper left corner */
      if(gp = tgetstr("G3", &tp))
       L = *gp;                /* Lower left corner */
      if(gp = tgetstr("GH", &tp))
       H = *gp;                /* Horizontal bar */
      if(gp = tgetstr("GR", &tp))
       T = *gp;                /* Left tee */
      if(gp = tgetstr("GV", &tp))
       V = *gp;                /* Vertical bar */
    }
# endif /* XENIX */
#endif  /* AIX */
    signal(SIGINT,  catchsig);
    signal(SIGQUIT, catchsig);
    signal(SIGTERM, catchsig);
  }

} /* initterm() */

/*
 *      Select graphical character set for device d
 */
static int initgraf(d)
  char *d;
{
  switch(*d) {
    default:                    /* Bad device */
      return(0);
    case 'a':                   /* ASCII meta characters */
      grafset = asciiset;
      break;
    case '8':                   /* IBM PC850 character set */
      grafset = ibmset;
      break;
    case 'p':                   /* Postscript */
      grafset = psset;
      ++isps;
      break;
    case 't':
      initterm(0);              /* Terminal */
      break;
  }
  return(1);

} /* initgraf() */

/*
 *      Write out graphic character c from current character set
 */
static void putgraf(c)
  int c;
{
  switch(c) {
    case 'I':                   /* Indention */
      putchar(I); break;
    case 'S':                   /* Space */
      putchar(S); break;
    default:                    /* Graphical characters */
      ACSCON;
      switch(c) {
       case 'H':               /* Horizontal bar */
         putchar(H); break;
       case 'L':               /* Lower left corner */
         putchar(L); break;
       case 'T':               /* Left tee */
         putchar(T); break;
       case 'U':               /* Upper left corner */
         putchar(U); break;
       case 'V':               /* Vertical bar */
         putchar(V); break;
      }
      ACSCOFF;
      break;
  }
  if(isps)
    putchar(' ');

} /* putgraf() */

/*
 *      Write header lines
 */
static void putheader()
{
  if(isps) {
    printf("%%!PS-Adobe-2.0\n");
    printf("%%%%Creator: %s\n", &sccsid[5]);
    printf("%%%%Title: %s: %s\n", title, root);
    printf("%%%%DocumentFont: %s\n", pft);
    printf("%%%%Pages: (atend)\n");
    printf("%% Copyright (c) 1991 by Peter Klingebiel & UNIX Magazin Muenchen\n");
    printf("%%%%EndComments\n");
    printf("%% Variables for font, fontsize, vertical space and indentions\n");
    printf("/ft /%s def\n", pft);
    printf("/ps %3.2lf def\n", pps);
    printf("/vs %3.2lf def\n", pvs);
    printf("/s1 %3.2lf def\n", pvs / 6.0);
    printf("/s2 %3.2lf def\n", pvs / 3.0);
    printf("/s3 %3.2lf def\n", pvs * 2.0 / 3.0);
    printf("%% Procedures for drawing and printing\n");
    if(rflag) {
      printf("/Z { gsave 0.95 setgray\n");
      printf("     0 ps s3 add     rlineto  %3.2lf 0 rlineto\n", PSWID);
      printf("     0 ps s3 add neg rlineto -%3.2lf 0 rlineto\n", PSWID);
      printf("     fill grestore s2 s2 rmoveto\n");
      printf("     gsave show grestore %3.2lf setlinewidth } bind def\n", pps / 20.0);
    }
    else
      printf("/Z { %3.2lf setlinewidth } bind def\n", pps / 20.0);
    printf("/F { findfont exch scalefont setfont } bind def\n");
    printf("/P { gsave s1 s1 rmoveto show grestore stroke } bind def\n");
    printf("/G { moveto } bind def\n");
    printf("/U { 0 s2 rlineto s2 0 rlineto 0 s2 neg rmoveto } bind def\n");
    printf("/L { 0 vs rmoveto 0 s3 neg rlineto s2 0 rlineto 0 s2 neg rmoveto } bind def\n");
    printf("/V { 0 vs rlineto s2 vs neg rmoveto } bind def\n");
    printf("/T { 0 vs rlineto 0 s3 neg rmoveto s2 0 rlineto 0 s2 neg rmoveto } bind def\n");
    printf("/S { s2 0 rmoveto } bind def\n");
    printf("/I { s2 %d mul 0 rmoveto } bind def\n", indent);
    printf("/H { 0 s2 rmoveto s2 %d mul 0 rlineto 0 s2 neg rmoveto } bind def\n", indent-1);
    printf("%%%%EndProlog\n");
    printf("ps ft F\n");
  }
  else if(rflag)
    printf("\n%s  %s: %s\n\n", pvers, title, root);

} /* putheader() */

/*
 *      Write trailing lines
 */
static void puttrailer()
{
  if(isps) {
    printf("showpage\n");
    printf("%%%%Trailer\n");
    printf("%%%%Pages: %d\n", psp);
    putchar(PSEOF);
  }

} /* puttrailer() */

/*
 *      Scan one entry line and write out
 */
static void putentry(s)
  char *s;
{
  int c, i;

  if(isps) {
    if(psp == 0 || psy < PSMIN) {
      if(psp)
       printf("showpage\n");
      printf("%%%%Page %d\n", ++psp);
      if(rflag) {
       printf("%3.2lf %3.2lf G (%s  %s    Page %d) Z\n", PSIND, PSMAX, title, root, psp);
       psy = PSMAX - 2.0 * pvs;
      }
      else {
       printf("Z\n");
       psy = PSMAX;
      }
    }
    printf("%3.2lf %3.2lf G ", PSIND, psy);
    psy -= pvs;
  }
  else
    putgraf(*s++);              /* First column */
  while(c = *s++)
    switch(c) {
      default:                  /* Bad format */
       return;
      case 'H':                 /* Horizontal bar + file name */
       if(isps) {
         putgraf(c);
         printf("(%s) P\n", s);
       }
       else {
         for(i = 1; i < indent; i++)
           putgraf(c);
         printf(" %s\n", s);
       }
       return;
      case 'I':                 /* Indention */
       if(isps)
         putgraf(c);
       else
         for(i = 0; i < indent; i++)
           putgraf(c);
       break;
      case 'S':                 /* One space */
      case 'L':                 /* Lower left corner */
      case 'T':                 /* Left tee */
      case 'U':                 /* Upper left corner */
      case 'V':                 /* Vertical bar */
       putgraf(c);
       break;
    }

} /* putentry() */

/*
 *      Main routine
 */
main(argc, argv)
  int argc;
  char **argv;
{
  char line[MAXLEN], *list;
  FILE *fp;
  int c;
  extern char *optarg;          /* From getopt() */
  extern int optind;

  /* Get programs name */
  if(pname = strrchr(argv[0], '/'))
    ++pname;
  else
    pname = argv[0];

  /* Parse and check command line options */
  while((c = getopt(argc, argv, OPTIONS)) != EOF) {
    switch(c) {
      default:                  /* Bad option */
       fprintf(stderr, "%s\n", &sccsid[5]);
       exit(1);
      case 'V':                 /* Version */
       fprintf(stderr, "%s\n", &sccsid[5]);
       exit(0);
      case 'h':                 /* Some help */
       fprintf(stderr, "%s\n", &usage[5]);
       fprintf(stderr, "\t-T dev\t\tOutput for device dev\n");
       fprintf(stderr, "\t-V\t\tDisplay program version\n");
       fprintf(stderr, "\t-d dev\t\tOutput for device dev\n");
       fprintf(stderr, "\t-f fnt\t\tUse font fnt (ps only, default: Courier-Bold)\n");
       fprintf(stderr, "\t-h\t\tDisplay some help\n");
       fprintf(stderr, "\t-i ind\t\tSet tree indention to ind (3..9)\n");
       fprintf(stderr, "\t-s siz\t\tUse font size siz (ps only, default: 10)\n");
       fprintf(stderr, "\t\t\tCurrently supported output devices are:\n");
       fprintf(stderr, "\t\t\tascii\tAscii meta graphic characters\n");
       fprintf(stderr, "\t\t\t850\tPrinters using IBM PC850 character set\n");
       fprintf(stderr, "\t\t\tps\tPostscript\n");
       fprintf(stderr, "\t\t\tterm\tTerminal (default)\n");
       exit(0);
      case 'T':                 /* Device for graphical character set */
      case 'd':
       if(initgraf(optarg))
         break;
       fprintf(stderr, "%s: device not supported -- %s\n", pname, optarg);
       exit(1);
      case 'r':                 /* Suppress pathname of rootdirectory */
       rflag = 0;
       break;
      case 'i':                 /* Indention */
       if((indent = atoi(optarg)) < MININD || indent > MAXIND) {
         fprintf(stderr, "%s: bad indention -- %d\n", pname, indent);
         exit(1);
       }
       break;
     case 'f':                  /* Postscript font */
       pft = optarg;
       break;
     case 's':                  /* Font size */
       pps = atof(optarg);
       pvs = pps + 2.0;
       break;
    }
  }

  /* Get and tree list file name */
  if((argc - optind) != 1) {
    fprintf(stderr, "%s\n", &usage[5]);
    exit(1);
  }
  else
    list = argv[optind];

  /* Open directory tree list file */
  if( !(fp = fopen(list, "r"))) {
    fprintf(stderr, "%s: Cannot open list file %s\n", pname, list);
    exit(1);
  }

  /* Init default graphical character set if needed */
  if(grafset == UNULL)
    initterm(1);

  /* Read, scan and write tree list file line by line */
  while(fgets(line, sizeof(line), fp)) {
    line[strlen(line)-1] = '\0';
    switch(c = line[0]) {
      default:                  /* Ignore */
       break;
      case 'R':                 /* Pathname of rootdirectory */
       strncpy(root, &line[1], sizeof(root));
       putheader();
       break;
      case 'H':                 /* Valid tree list entry */
      case 'L':
      case 'U':
      case 'V':
       putentry(line);
       break;
    }
  }
  fclose(fp);

  /* Write trailer and exit */
  puttrailer();
  exit(0);

} /* main() */

